File: backend/apps/internships/filters.py
import django_filters
from django.db.models import Q

from .models import Internship


class InternshipFilter(django_filters.FilterSet):
    """
    Filters for active and verified internships.
    """

    country = django_filters.CharFilter(
        field_name="country",
        lookup_expr="icontains",
    )

    city = django_filters.CharFilter(
        field_name="city",
        lookup_expr="icontains",
    )

    location = django_filters.CharFilter(
        method="filter_location",
    )

    category = django_filters.CharFilter(
        field_name="category",
        lookup_expr="icontains",
    )

    type = django_filters.ChoiceFilter(
        field_name="internship_type",
        choices=Internship.INTERNSHIP_TYPE_CHOICES,
    )

    work_mode = django_filters.ChoiceFilter(
        field_name="work_mode",
        choices=Internship._meta.get_field("work_mode").choices,
    )

    work_type = django_filters.ChoiceFilter(
        choices=Internship.WORK_TYPE_CHOICES,
    )

    experience = django_filters.CharFilter(
        field_name="required_experience",
        lookup_expr="icontains",
    )

    internship_type = django_filters.ChoiceFilter(
        choices=Internship.INTERNSHIP_TYPE_CHOICES,
    )

    compensation_type = django_filters.ChoiceFilter(
        choices=Internship.COMPENSATION_TYPE_CHOICES,
    )

    minimum_compensation = django_filters.NumberFilter(
        field_name="minimum_compensation",
        lookup_expr="gte",
    )

    maximum_compensation = django_filters.NumberFilter(
        field_name="maximum_compensation",
        lookup_expr="lte",
    )

    duration_min_weeks = django_filters.NumberFilter(
        field_name="duration_min_weeks",
        lookup_expr="gte",
    )

    duration_max_weeks = django_filters.NumberFilter(
        field_name="duration_max_weeks",
        lookup_expr="lte",
    )

    skill = django_filters.CharFilter(
        method="filter_skill",
    )

    def filter_location(self, queryset, name, value):
        lookup = Q(country__icontains=value) | Q(
            city__icontains=value) | Q(location_text__icontains=value)
        return queryset.filter(lookup)

    def filter_skill(self, queryset, name, value):
        """
        Filter internships whose required_skills contain
        the requested skill.
        """

        return queryset.filter(
            required_skills__icontains=value
        )

    class Meta:
        model = Internship

        fields = [
            "country",
            "city",
            "location",
            "category",
            "type",
            "work_mode",
            "internship_type",
            "work_type",
            "experience",
            "compensation_type",
            "minimum_compensation",
            "maximum_compensation",
            "duration_min_weeks",
            "duration_max_weeks",
        ]

File: frontend/src/types/filters.ts
/**
 * Filter types for internship search
 */

export interface InternshipFilters {
  search?: string
  country?: string
  city?: string
  location?: string
  category?: string
  internship_type?: 'remote' | 'onsite' | 'hybrid'
  work_mode?: 'remote' | 'onsite' | 'hybrid' | 'any'
  work_type?: 'full_time' | 'part_time'
  compensation_type?: 'paid' | 'unpaid' | 'unknown'
  minimum_compensation?: number
  maximum_compensation?: number
  duration_min_weeks?: number
  duration_max_weeks?: number
  skill?: string
}

export const INTERNSHIP_TYPE_CHOICES = [
  { value: 'remote', label: 'Remote' },
  { value: 'onsite', label: 'On-site' },
  { value: 'hybrid', label: 'Hybrid' },
] as const

export const WORK_MODE_CHOICES = [
  { value: 'remote', label: 'Remote' },
  { value: 'onsite', label: 'On-site' },
  { value: 'hybrid', label: 'Hybrid' },
  { value: 'any', label: 'Any' },
] as const

export const WORK_TYPE_CHOICES = [
  { value: 'full_time', label: 'Full-time' },
  { value: 'part_time', label: 'Part-time' },
] as const

export const COMPENSATION_TYPE_CHOICES = [
  { value: 'paid', label: 'Paid' },
  { value: 'unpaid', label: 'Unpaid' },
  { value: 'unknown', label: 'Unknown' },
] as const
